libpod/image/image.go

```diff
@@ -210,82 +194,208 @@ func (ir *Runtime) SaveImages(ctx context.Context, namesOrIDs []string, format s
+			if err != nil {
+				return errors.Wrapf(err, "error getting source imageReference for %q", img.InputName)
+			}
+			_, err = cp.Image(ctx, policyContext, dest, src, copyOptions)
+			if err != nil {
+				return err
+			}
+		}
+	case OCIArchive:
+		archWriter, err := ociarchive.NewWriter(sys, outputFile)
 		if err != nil {
 			return err
 		}
+		defer func() {
+			err := archWriter.Close()
+			if err == nil {
+				return
+			}
+			if finalErr == nil {
+				finalErr = err
+			}
+			finalErr = errors.Wrap(finalErr, err.Error())
+		}()
 
-		img := imageMap[id]
-		copyOptions := getCopyOptions(sys, writer, nil, nil, SigningOptions{RemoveSignatures: removeSignatures}, "", img.tags)
+		copyOptions := getCopyOptions(sys, writer, nil, nil, SigningOptions{RemoveSignatures: removeSignatures}, "", nil)
 		copyOptions.DestinationCtx.SystemRegistriesConfPath = registries.SystemRegistriesConfPath()
+		// copy the image one by one
+		for _, nameOrID := range namesOrIDs {
+			localImage, err := ir.NewFromLocal(nameOrID)
+			if err != nil {
+				return err
+			}
+			id := localImage.ID()
+			// For copying, we need a source reference that we can create
+			// from the image.
+			src, err := is.Transport.NewStoreReference(localImage.imageruntime.store, nil, id)
+			if err != nil {
+				return errors.Wrapf(err, "error getting source imageReference for %q", nameOrID)
+			}
+			// if the nameOrID referred to an ID, save all the tagged images with same ID
+			// otherwise only save image with the inputname
+			if len(localImage.Names()) > 0 && strings.HasPrefix(id, nameOrID) {
+				for _, name := range localImage.Names() {
+					dest, err := archWriter.NewReference(name)
+					if err != nil {
+						return err
+					}
+					_, err = cp.Image(ctx, policyContext, dest, src, copyOptions)
+					if err != nil {
+						return err
+					}
+				}
+				continue
+			}
+			var dest types.ImageReference
+			if strings.HasPrefix(id, nameOrID) {
+				// the nameOrID referred to an ID but the image has no name
+				dest, err = archWriter.NewReference("")
+				if err != nil {
+					return err
+				}
+			} else {
+				// the nameOrID referred to a name
+				dest, err = archWriter.NewReference(nameOrID)
+				if err != nil {
+					return err
+				}
+			}
+			_, err = cp.Image(ctx, policyContext, dest, src, copyOptions)
+			if err != nil {
+				return err
+			}
+		}
+	}
 
-		// For copying, we need a source reference that we can create
-		// from the image.
-		src, err := is.Transport.NewStoreReference(img.imageruntime.store, nil, id)
-		if err != nil {
-			return errors.Wrapf(err, "error getting source imageReference for %q", img.InputName)
+	return nil
+}
+
+// LoadAllImagesFromOCIArchive loads all images from the docker archive that
+// fileName points to.
+func (ir *Runtime) LoadAllImagesFromOCIArchive(ctx context.Context, fileName string, signaturePolicyPath string, writer io.Writer) ([]string, error) {
+	sc := GetSystemContext(signaturePolicyPath, "", false)
```

Comment:
`LoadAllImagesFromDockerArchive` has an `if signaturePolicyPath == ""` check. Which one is right?